homeassistant/components/withings/sensor.py

```diff
@@ -645,9 +644,32 @@ def _async_add_activity_entities() -> None:
 
     sleep_coordinator = withings_data.sleep_coordinator
 
-    entities.extend(
-        WithingsSleepSensor(sleep_coordinator, attribute) for attribute in SLEEP_SENSORS
+    sleep_callback: Callable[[], None] | None = None
+
+    sleep_entities_setup_before = ent_reg.async_get_entity_id(
+        Platform.SENSOR,
+        DOMAIN,
+        f"withings_{entry.unique_id}_sleep_deep_duration_seconds",
     )
+
+    def _async_add_sleep_entities() -> None:
+        """Add sleep entities."""
+        if sleep_coordinator.data is not None:
+            async_add_entities(
+                WithingsSleepSensor(sleep_coordinator, attribute)
+                for attribute in SLEEP_SENSORS
+            )
+            if sleep_callback:
+                sleep_callback()
+
+    if sleep_coordinator.data is not None or sleep_entities_setup_before:
+        entities.extend(
+            WithingsSleepSensor(sleep_coordinator, attribute)
+            for attribute in SLEEP_SENSORS
+        )
+    else:
+        sleep_callback = sleep_coordinator.async_add_listener(_async_add_sleep_entities)
```

Comment:
```suggestion
        remove_listener: Callable[[], None]

        def _async_add_sleep_entities() -> None:
            """Add sleep entities."""
            if sleep_coordinator.data is not None:
                async_add_entities(
                    WithingsSleepSensor(sleep_coordinator, attribute)
                    for attribute in SLEEP_SENSORS
                )
                remove_listener()

        remove_listener = sleep_coordinator.async_add_listener(
            _async_add_sleep_entities
        )
```

I renamed `sleep_callback` to `remove_listener` for clarity.
I moved the callback function inside the else block so I can remove `| None = None` from the listener variable